lib/active_storage_validations/aspect_ratio_validator.rb

```diff
@@ -1,120 +1,149 @@
-
-      return if x > 0 && y > 0 && (x.to_f / y).round(PRECISION) == (metadata[:width].to_f / metadata[:height]).round(PRECISION)
-
-      errors_options = initialize_error_options(options, attachable)
-      errors_options[:aspect_ratio] = "#{x}:#{y}"
-      add_error(record, attribute, :aspect_ratio_is_not, **errors_options)
-    end
-
-    def ensure_at_least_one_validator_option
-      unless AVAILABLE_CHECKS.any? { |argument| options.key?(argument) }
-        raise ArgumentError, 'You must pass :with to the validator'
-      end
-    end
-
-    def ensure_aspect_ratio_validity
-      return true if options[:with]&.is_a?(Proc)
-
-      unless NAMED_ASPECT_RATIOS.include?(options[:with]) || options[:with] =~ ASPECT_RATIO_REGEX
-        raise ArgumentError, <<~ERROR_MESSAGE
-          You must pass a valid aspect ratio to the validator
-          It should either be a named aspect ratio (#{NAMED_ASPECT_RATIOS.join(', ')})
-          Or an aspect ratio like 'is_16_9' (matching /#{ASPECT_RATIO_REGEX.source}/)
-        ERROR_MESSAGE
-      end
-    end
-  end
-end
+# frozen_string_literal: true
+
+require_relative 'shared/asv_active_storageable'
+require_relative 'shared/asv_analyzable'
+require_relative 'shared/asv_attachable'
+require_relative 'shared/asv_errorable'
+require_relative 'shared/asv_optionable'
+require_relative 'shared/asv_symbolizable'
+
+module ActiveStorageValidations
+  class AspectRatioValidator < ActiveModel::EachValidator # :nodoc
+    include ASVActiveStorageable
+    include ASVAnalyzable
+    include ASVAttachable
+    include ASVErrorable
+    include ASVOptionable
+    include ASVSymbolizable
+
+    AVAILABLE_CHECKS = %i[with in].freeze
+    NAMED_ASPECT_RATIOS = %i[square portrait landscape].freeze
+    ASPECT_RATIO_REGEX = /is_([1-9]\d*)_([1-9]\d*)/.freeze
+    ERROR_TYPES = %i[
+      aspect_ratio_not_square
+      aspect_ratio_not_portrait
+      aspect_ratio_not_landscape
+      aspect_ratio_is_not
+      aspect_ratio_invalid
+      image_metadata_missing
+    ].freeze
+    PRECISION = 3.freeze
+
+    def check_validity!
+      ensure_at_least_one_validator_option
+      ensure_aspect_ratio_validity
+    end
+
+    def validate_each(record, attribute, _value)
+      return if no_attachments?(record, attribute)
+
+      validate_changed_files_from_metadata(record, attribute)
+    end
+
+    private
+
+    def is_valid?(record, attribute, attachable, metadata)
+      flat_options = set_flat_options(record)
+
+      return if image_metadata_missing?(record, attribute, attachable, flat_options, metadata)
+
+      @authorized_aspect_ratios = authorized_aspect_ratios(flat_options).compact
+      
+      return true if authorized_aspect_ratio?(record, attribute, attachable, metadata)
+
+      errors_options = initialize_error_options(options, attachable)
+      errors_options[:aspect_ratio] = string_aspect_ratios(flat_options)
+      add_error(record, attribute, aspect_ratio_error_mapping, **errors_options)
+      false
```

Comment:
As previously stated, I would suggest doing the same as `content_type_validator` for the sake of clarity.

```ruby
    def is_valid?(record, attribute, attachable, metadata)
      authorized_aspect_ratio?(record, attribute, attachable, metadata)
    end
```

And then do the below logic in the `authorized_aspect_ratio?` method.